config/src/main/java/com/typesafe/config/impl/ConfigImpl.java

```diff
@@ -360,6 +361,57 @@ public static void reloadEnvVariablesConfig() {
         EnvVariablesHolder.envVariables = loadEnvVariables();
     }
 
+    private static AbstractConfigObject loadEnvVariablesOverrides() {
+        Map<String, String> env = new HashMap(System.getenv());
+        Map<String, String> result = new HashMap(System.getenv());
+        for (String key : env.keySet()) {
+            if (key.startsWith(ENV_VAR_OVERRIDE_PREFIX)) {
+                StringBuilder builder = new StringBuilder();
+
+                String strippedPrefix = key.substring(ENV_VAR_OVERRIDE_PREFIX.length(), key.length());
+
+                int underscores = 0;
+                for (char c : strippedPrefix.toCharArray()) {
+                    if (c == '_') {
+                        underscores++;
+                    } else {
+                        switch (underscores) {
+                            case 1: builder.append('.');
+                                    break;
+                            case 2: builder.append('-');
+                                    break;
+                            case 3: builder.append('_');
+                                    break;
+                        }
+                        underscores = 0;
+                        builder.append(c);
+                    }
+                }
+
+                String propertyKey = builder.toString();
+                result.put(propertyKey, env.get(key));
+            }
+        }
+
+        return PropertiesParser.fromStringMap(newSimpleOrigin("env variables overrides"), result);
+    }
+
+    private static class EnvVariablesOverridesHolder {
+        static volatile AbstractConfigObject envVariables = loadEnvVariablesOverrides();
```

Comment:
I guess we need to make ConfigFactory.invalidateCaches clear this https://github.com/lightbend/config/blob/64c60f071354b46fcf433e788f2360bfc8797db5/config/src/main/java/com/typesafe/config/ConfigFactory.java#L495

Reply:
done